### ISSUE TITLE
ansible_runner error with ansible 2.8

### ISSUE THREAD

**Issue by dsavineau** [philpep/testinfra#437]
Trying the latest testinfra release (3.0.0) with ansible 2.8 (rc2) but running into this issue:

```console
dev-centos-non_container-lvm_osds run-test: commands[7] | py.test -n 8 --durations=0 --sudo -v --connection=ansible --ansible-inventory=/home/jenkins-build/build/workspace/ceph-ansible-prs-dev-centos-non_container-lvm_osds/tests/functional/lvm-osds/hosts /home/jenkins-build/build/workspace/ceph-ansible-prs-dev-centos-non_container-lvm_osds/tests/functional/tests
[2211] /home/jenkins-build/build/workspace/ceph-ansible-prs-dev-centos-non_container-lvm_osds/tests/functional/lvm-osds$ /tmp/venv.YMrKxhXs5g/dev-centos-non_container-lvm_osds/bin/py.test -n 8 --durations=0 --sudo -v --connection=ansible --ansible-inventory=/home/jenkins-build/build/workspace/ceph-ansible-prs-dev-centos-non_container-lvm_osds/tests/functional/lvm-osds/hosts /home/jenkins-build/build/workspace/ceph-ansible-prs-dev-centos-non_container-lvm_osds/tests/functional/tests
============================= test session starts ==============================
platform linux2 -- Python 2.7.5, pytest-4.4.1, py-1.8.0, pluggy-0.9.0 -- /tmp/venv.YMrKxhXs5g/dev-centos-non_container-lvm_osds/bin/python
cachedir: /tmp/venv.YMrKxhXs5g/dev-centos-non_container-lvm_osds/.pytest_cache
rootdir: /home/jenkins-build/build/workspace/ceph-ansible-prs-dev-centos-non_container-lvm_osds/tests, inifile: pytest.ini
plugins: testinfra-3.0.0, xdist-1.28.0, forked-1.0.2
(...)
==================================== ERRORS ====================================
______________ ERROR collecting functional/tests/test_install.py _______________
/tmp/venv.YMrKxhXs5g/dev-centos-non_container-lvm_osds/lib/python2.7/site-packages/pluggy/hooks.py:289: in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
/tmp/venv.YMrKxhXs5g/dev-centos-non_container-lvm_osds/lib/python2.7/site-packages/pluggy/manager.py:68: in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
/tmp/venv.YMrKxhXs5g/dev-centos-non_container-lvm_osds/lib/python2.7/site-packages/pluggy/manager.py:62: in <lambda>
    firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
/tmp/venv.YMrKxhXs5g/dev-centos-non_container-lvm_osds/lib/python2.7/site-packages/_pytest/python.py:238: in pytest_pycollect_makeitem
    res = list(collector._genfunctions(name, obj))
/tmp/venv.YMrKxhXs5g/dev-centos-non_container-lvm_osds/lib/python2.7/site-packages/_pytest/python.py:414: in _genfunctions
    self.ihook.pytest_generate_tests(metafunc=metafunc)
/tmp/venv.YMrKxhXs5g/dev-centos-non_container-lvm_osds/lib/python2.7/site-packages/pluggy/hooks.py:289: in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
/tmp/venv.YMrKxhXs5g/dev-centos-non_container-lvm_osds/lib/python2.7/site-packages/pluggy/manager.py:68: in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
/tmp/venv.YMrKxhXs5g/dev-centos-non_container-lvm_osds/lib/python2.7/site-packages/pluggy/manager.py:62: in <lambda>
    firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
/tmp/venv.YMrKxhXs5g/dev-centos-non_container-lvm_osds/lib/python2.7/site-packages/testinfra/plugin.py:113: in pytest_generate_tests
    ansible_inventory=metafunc.config.option.ansible_inventory,
/tmp/venv.YMrKxhXs5g/dev-centos-non_container-lvm_osds/lib/python2.7/site-packages/testinfra/host.py:143: in get_hosts
    for backend in testinfra.backend.get_backends(hosts, **kwargs):
/tmp/venv.YMrKxhXs5g/dev-centos-non_container-lvm_osds/lib/python2.7/site-packages/testinfra/backend/__init__.py:91: in get_backends
    for name in klass.get_hosts(host, **kw):
/tmp/venv.YMrKxhXs5g/dev-centos-non_container-lvm_osds/lib/python2.7/site-packages/testinfra/backend/ansible.py:59: in get_hosts
    return AnsibleRunner.get_runner(inventory).get_hosts(host)
/tmp/venv.YMrKxhXs5g/dev-centos-non_container-lvm_osds/lib/python2.7/site-packages/testinfra/utils/ansible_runner.py:124: in get_hosts
    groupmatch = fnmatch.fnmatch(group, pattern)
/tmp/venv.YMrKxhXs5g/dev-centos-non_container-lvm_osds/lib64/python2.7/fnmatch.py:43: in fnmatch
    return fnmatchcase(name, pat)
/tmp/venv.YMrKxhXs5g/dev-centos-non_container-lvm_osds/lib64/python2.7/fnmatch.py:75: in fnmatchcase
    res = translate(pat)
/tmp/venv.YMrKxhXs5g/dev-centos-non_container-lvm_osds/lib64/python2.7/fnmatch.py:87: in translate
    i, n = 0, len(pat)
E   TypeError: object of type 'NoneType' has no len()
```
I also test 3.0.0 release with ansible 2.7 but the result is the same.

The ansible inventory file is really simple:
```yaml
[mons]
mon0

[osds]
osd0
```

**Comment by philpep** [comment]
Hi, thanks for reporting ! Indeed this is a regression since 3.0.0, you can workaround the bug by replacing ``--connection=ansible`` by ``--hosts ansible://all`` or just wait the 3.0.1 bugfix version which will come in a few hours.

**Comment by philpep** [comment]
This is fixed in released 3.0.1

**Comment by dsavineau** [comment]
Thanks for the fix !
Unfortunately I've a different issue now with 3.0.1. I didn't test using `--hosts ansible://all` yet but I will give it a try.

```console
dev-centos-container-lvm_osds run-test: commands[7] | py.test -n 8 --durations=0 --sudo -v --connection=ansible --ansible-inventory=/home/jenkins-build/build/workspace/ceph-ansible-prs-dev-centos-container-lvm_osds/tests/functional/lvm-osds/container/hosts /home/jenkins-build/build/workspace/ceph-ansible-prs-dev-centos-container-lvm_osds/tests/functional/tests
[21624] /home/jenkins-build/build/workspace/ceph-ansible-prs-dev-centos-container-lvm_osds/tests/functional/lvm-osds/container$ /tmp/venv.cxgARmg8wG/dev-centos-container-lvm_osds/bin/py.test -n 8 --durations=0 --sudo -v --connection=ansible --ansible-inventory=/home/jenkins-build/build/workspace/ceph-ansible-prs-dev-centos-container-lvm_osds/tests/functional/lvm-osds/container/hosts /home/jenkins-build/build/workspace/ceph-ansible-prs-dev-centos-container-lvm_osds/tests/functional/tests
============================= test session starts ==============================
platform linux2 -- Python 2.7.5, pytest-4.4.1, py-1.8.0, pluggy-0.9.0 -- /tmp/venv.cxgARmg8wG/dev-centos-container-lvm_osds/bin/python
cachedir: /tmp/venv.cxgARmg8wG/dev-centos-container-lvm_osds/.pytest_cache
rootdir: /home/jenkins-build/build/workspace/ceph-ansible-prs-dev-centos-container-lvm_osds/tests, inifile: pytest.ini
plugins: testinfra-3.0.1, xdist-1.28.0, forked-1.0.2
(...)
______________ ERROR collecting functional/tests/test_install.py _______________
/tmp/venv.cxgARmg8wG/dev-centos-container-lvm_osds/lib/python2.7/site-packages/pluggy/hooks.py:289: in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
/tmp/venv.cxgARmg8wG/dev-centos-container-lvm_osds/lib/python2.7/site-packages/pluggy/manager.py:68: in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
/tmp/venv.cxgARmg8wG/dev-centos-container-lvm_osds/lib/python2.7/site-packages/pluggy/manager.py:62: in <lambda>
    firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
/tmp/venv.cxgARmg8wG/dev-centos-container-lvm_osds/lib/python2.7/site-packages/_pytest/python.py:238: in pytest_pycollect_makeitem
    res = list(collector._genfunctions(name, obj))
/tmp/venv.cxgARmg8wG/dev-centos-container-lvm_osds/lib/python2.7/site-packages/_pytest/python.py:414: in _genfunctions
    self.ihook.pytest_generate_tests(metafunc=metafunc)
/tmp/venv.cxgARmg8wG/dev-centos-container-lvm_osds/lib/python2.7/site-packages/pluggy/hooks.py:289: in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
/tmp/venv.cxgARmg8wG/dev-centos-container-lvm_osds/lib/python2.7/site-packages/pluggy/manager.py:68: in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
/tmp/venv.cxgARmg8wG/dev-centos-container-lvm_osds/lib/python2.7/site-packages/pluggy/manager.py:62: in <lambda>
    firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
/tmp/venv.cxgARmg8wG/dev-centos-container-lvm_osds/lib/python2.7/site-packages/testinfra/plugin.py:113: in pytest_generate_tests
    ansible_inventory=metafunc.config.option.ansible_inventory,
/tmp/venv.cxgARmg8wG/dev-centos-container-lvm_osds/lib/python2.7/site-packages/testinfra/host.py:143: in get_hosts
    for backend in testinfra.backend.get_backends(hosts, **kwargs):
/tmp/venv.cxgARmg8wG/dev-centos-container-lvm_osds/lib/python2.7/site-packages/testinfra/backend/__init__.py:91: in get_backends
    for name in klass.get_hosts(host, **kw):
/tmp/venv.cxgARmg8wG/dev-centos-container-lvm_osds/lib/python2.7/site-packages/testinfra/backend/ansible.py:59: in get_hosts
    return AnsibleRunner.get_runner(inventory).get_hosts(host or "all")
/tmp/venv.cxgARmg8wG/dev-centos-container-lvm_osds/lib/python2.7/site-packages/testinfra/utils/ansible_runner.py:130: in get_hosts
    result |= set(itergroup(group))
/tmp/venv.cxgARmg8wG/dev-centos-container-lvm_osds/lib/python2.7/site-packages/testinfra/utils/ansible_runner.py:120: in itergroup
    for host in itergroup(g):
/tmp/venv.cxgARmg8wG/dev-centos-container-lvm_osds/lib/python2.7/site-packages/testinfra/utils/ansible_runner.py:117: in itergroup
    for host in inventory[group].get('hosts', []):
E   KeyError: u'ungrouped'
```

Do you prefer to keep tracking the issues with 3.0.x there or should I create a different one ?

**Comment by philpep** [comment]
Hi @dsavineau , thanks for reporting this one too, It should be fixed by https://github.com/philpep/testinfra/pull/440 which has been released in 3.0.2
